### DISCUSSION TITLE
How to speed up "$_com1 = "virtual" in .dosemurc?

### DISCUSSION THREAD

**Discussion by Baron-von-Riedesel** [dosemu2/dosemu2#1698] [General]
It works, but is pretty slow - changing the baud rate inside DOS has no effect.

Btw.:

>  "virtual" keyword may be used to redirect the COM port to the terminal.
>  A wrong but quick way for the sysop to play his BBS door games locally.

What is the "correct" way to redirect COM1 to the linux console?

**Comment by stsp** [comment]
> pretty slow

Too vague unless you provide
a test-case.

> changing the baud rate inside DOS has no effect.

Try `speed 0`, maybe the i/o
gets throttled.

> What is the "correct" way to redirect COM1 to the linux console?

By using the DOS terminal emulator?
This "virtual" flag just hacks up one for
you.

**Reply by stsp** [reply]
Sorry, not a terminal emulator,
but a com port redirector:
https://www.bbscorner.com/telnetbbs/virtualmodems.htm
But in this case it should also
be TSR, so I wonder if such
exist at all?
If not, I should correct the
comment about "virtual"
keyword in dosemu.conf.

What do you think?

**Reply by stsp** [reply]
Also this "virtual" thing works
only in terminal mode (-t) and
is completely ignored for gfx
modes.
But its probably OK for such
a hack... at least no one ever
complained.

**Reply by stsp** [reply]
Or actually it works with any
mode, it just doesn't shut down
local output for non-terminal modes...
Cool, should be properly documented.

**Reply by stsp** [reply]
OK, actually, what's the use?
Isn't it the same as just to set
`$_com1="/dev/tty"`?
Yes, for `-t` mode it also shuts
down screen output, but who
needs that, just don't use `-t`
when redirecting com port to
/dev/tty?

**Comment by ecm-pushbx** [comment]
I use the following setup:

1. Run `socat pty,link=/tmp/vptty-dos,rawer pty,link=/tmp/vptty-linux,rawer & disown`
2. In the terminal you want to have the serial I/O in, run `picocom /tmp/vptty-linux | tee -a /tmp/trace.log`
3. Start dosemu2 with `-I "serial { com 2 device /tmp/vptty-dos }"`

In step 2 I redirect the picocom output into a file [for use with my traclist program](https://pushbx.org/ecm/doc/traclist.htm#setup).

In step 3 I usually use the COM2 port because PD ZModem and lDebug both default to using COM2 (and that's where I have my physical null modem cable connected to my DOS machine too).

**Reply by stsp** [reply]
> In step 2 I redirect the picocom output into a file

dosemu2 can write directly to
a file from emulated serial port,
IIRC. You don't need socat+picocom+bash_redirection
just to achieve that.

**Reply by ecm-pushbx** [reply]
> dosemu2 can write directly to a file from emulated serial port, IIRC. You don't need socat+picocom+bash_redirection just to achieve that.

I want the picocom to use as an interface to the port I/O though, to control my application. The file is just for an *additional* read access.

**Comment by Baron-von-Riedesel** [comment]
> Too vague unless you provide a test-case.

It was about 2400 baud.

But it has become clear what the problem is: my program accessed the UART directly - changing to Int 0x14 makes communication fast.

**Reply by stsp** [reply]
int14 is not doing anything
special. It deliberately controls
the port via its registers interface,
rather than to do some fancy stuff.
So have you tried `speed 0`?

**Comment by Baron-von-Riedesel** [comment]
> So have you tried speed 0?

Yes, but didn't notice a speed difference.

The program I'm playing with is win16 debugger, written almost 30 years ago, rather complex, so it's perhaps not a really good test case, because I'm not 100% sure that I know what the program is doing - perhaps there's some strange/silly "handshake" code somewhere...

It's this thing:  https://github.com/Baron-von-Riedesel/debxxf

The win16 version has the advantage that it runs in the very same client as the debuggee, so it's currently ( until that gcc compiler bug is fixed ) the only version that I can test under DosEmu...

**Reply by stsp** [reply]
What compiler bug?
We already fixed everything.

**Comment by Baron-von-Riedesel** [comment]
Well, that infamous "internal compiler error" bug that currently prevents me from compiling DosEmu...

**Reply by stsp** [reply]
But I already told you this:
`CC=clang ./default-configure`
or:
`./default-configure -d`

**Comment by Baron-von-Riedesel** [comment]
You did, yes, and I tried .... and it didn't work :)) [forgot what the problem was]

**Reply by stsp** [reply]
`make clean` was a problem.

**Comment by Baron-von-Riedesel** [comment]
Ok thanks!
the compile process now has changed to clang and finishes correctly.
Trying to start dosemu, I get: 

ERROR: /usr/local/lib/dosemu/libplugin_fdpp.so: undefined symbol: FdppKernelFree
ERROR: can't load fdpp
ERROR: PKT: Cannot open tap0: Operation not permitted
ERROR: PKT: Cannot open tap0: Operation not permitted
ERROR: Bootable drive not found, exiting

**Comment by Baron-von-Riedesel** [comment]
Ok, I recreated fdpp, and now dosemu starts...

**Reply by stsp** [reply]
Good.
Now could you please try
to set `$_com1 = "/dev/tty"`
instead of virtual, and tell
me if its the same thing?
I need to either properly
document this "virtual" or
remove, but so far I don't
quite understand its use-case.

**Comment by Baron-von-Riedesel** [comment]
A little side-note:

Dosemu starts, but did you already test that it behaves like the old gcc version?

Because, the first thing i tried was to launch DD ( the "dark debugger" ). It's not supposed to run under DosEmu, but its error msg has changed. With the dosemu gcc version, it displayed "already in protected mode", now with the clang version I'm told: "80x86: this program requires 80386+ cpu". Is this possible?

**Reply by stsp** [reply]
Easy to check.
Use `./default-configure -d`
instead of clang and let us know
the outcome.
Don't forget `make clean` this time.

**Comment by Baron-von-Riedesel** [comment]
> Now could you please try to set $_com1 = "/dev/tty" instead of virtual, and tell me if its the same thing?

Yes, seems to behave identical, AFAICS

**Reply by stsp** [reply]
Could you please give me a
small test-case so that I can
re-implement this thing?

**Comment by Baron-von-Riedesel** [comment]
It's indeed a different error msg displayed by the "dark debugger" - strange!

But back to the COM thing - when the line

$_com1 = "..."

is activated in .dosemurc, and DosEmu has terminated, the console is "confused" - it accepts keyboard input, but it isn't echoed anymore.
This happens regardless whether any DOS program has written to COM1 or not. I'm not sure, but AFAIR this used to work better ...

**Reply by stsp** [reply]
> $_com1 = "..."

Not "..." but "virtual".
It needs to be re-implemented.

**Comment by Baron-von-Riedesel** [comment]
> Could you please give me a small test-case so that I can re-implement this thing?

Doesn't
C:\>CTTY com1
suffice?

**Reply by stsp** [reply]
Suffice what?

**Comment by Baron-von-Riedesel** [comment]
> Suffice what?

suffice as a test case, of course.

**Reply by stsp** [reply]
I want a test-case of some
bbs game, or any prog that
writes and reads to a COM
port.
Then I can re-implement
"virtual" as probably a simple
"ctty com1". Will that work?

**Reply by stsp** [reply]
So what does your `ctty com1`
do if there is no one on a "remote"
side of a com port? I suppose
its not a test-case as it will not
read or write anything w/o commands
from remote.

**Reply by stsp** [reply]
It would be good if I could
connect comX and con, and
allow the bbs game to work
on a remote side of a com.
But I don't think there is a
socat port for DOS.
`ctty com1` stops listening to
con, which is not what I need.

**Comment by Baron-von-Riedesel** [comment]
> I suppose its not a test-case as it will not read or write anything w/o commands from remote.

It just redirects the DOS console to a COM port. So it will write "C:\>" to the linux terminal and accept commands from there. That should be enough for testing - but perhaps I'm misunderstanding you...

**Reply by stsp** [reply]
I want it to work from any
dosemu window, and from terminal -
only in `-t` mode.
That should be the key difference
with dosemu1 impl.

**Reply by stsp** [reply]
So what I want is to connect
comX<-->con. We already have
dos2tty subsystem in dosemu2
for that.

**Reply by stsp** [reply]
So the test-case that writes
some random crap to serial
port, and that does simple echo,
would be good to have.

**Reply by stsp** [reply]
In fact, I can use `echo aaa >com1`
in a bat file. So can write a test-case
myself.
